planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp

```diff
@@ -3506,34 +3498,113 @@ void AvoidanceModule::insertWaitPoint(
+    getShiftLength(object, utils::avoidance::isOnRight(object), avoid_margin));
   const auto constant =
     p->min_prepare_distance + object_parameter.safety_buffer_longitudinal + base_link2front;
-  const auto start_longitudinal =
-    o_front.longitudinal -
-    std::clamp(variable + constant, p->stop_min_distance, p->stop_max_distance);
+
+  return object.longitudinal -
+         std::clamp(variable + constant, p->stop_min_distance, p->stop_max_distance);
+}
+
+void AvoidanceModule::insertWaitPoint(
+  const bool use_constraints_for_decel, ShiftedPath & shifted_path) const
+{
+  const auto & data = avoidance_data_;
+
+  if (!data.stop_target_object) {
+    return;
+  }
+
+  if (data.avoiding_now) {
+    return;
+  }
 
   if (!use_constraints_for_decel) {
     utils::avoidance::insertDecelPoint(
-      getEgoPosition(), start_longitudinal, 0.0, shifted_path.path, stop_pose_);
+      getEgoPosition(), data.to_stop_line, 0.0, shifted_path.path, stop_pose_);
     return;
   }
 
   const auto stop_distance = getMildDecelDistance(0.0);
   if (stop_distance) {
-    const auto insert_distance = std::max(start_longitudinal, *stop_distance);
+    const auto insert_distance = std::max(data.to_stop_line, *stop_distance);
     utils::avoidance::insertDecelPoint(
       getEgoPosition(), insert_distance, 0.0, shifted_path.path, stop_pose_);
   }
 }
 
+void AvoidanceModule::insertStopPoint(
+  const bool use_constraints_for_decel, ShiftedPath & shifted_path) const
+{
+  const auto & data = avoidance_data_;
+
+  if (data.safe) {
+    return;
+  }
+
+  // step1. find index where the ego is within the current lane.
+  const auto stop_idx = [&]() {
+    const auto ego_idx = planner_data_->findEgoIndex(shifted_path.path.points);
+    for (size_t idx = ego_idx; idx < shifted_path.path.points.size(); ++idx) {
+      const auto & estimated_pose = shifted_path.path.points.at(idx).point.pose;
+      if (!utils::isEgoWithinOriginalLane(
+            data.current_lanelets, estimated_pose, planner_data_->parameters)) {
+        return idx - 1;
+      }
+    }
+
+    return shifted_path.path.points.size() - 2;
+  }();
+
+  // step2. calculate precise stop distance.
+  const auto stop_distance = [&]() {
+    const auto segment_distance =
+      calcSignedArcLength(shifted_path.path.points, stop_idx, stop_idx + 1);
+
+    constexpr double BUFFER = 1.0;
+    constexpr double dx = 0.1;
+    for (double x = dx; x < segment_distance + BUFFER; x += dx) {
+      const auto stop_pose = calcLongitudinalOffsetPose(shifted_path.path.points, stop_idx, x);
+
+      if (!stop_pose) {
+        continue;
+      }
+
+      if (!utils::isEgoWithinOriginalLane(
+            data.current_lanelets, stop_pose.get(), planner_data_->parameters)) {
+        return calcSignedArcLength(shifted_path.path.points, getEgoPosition(), stop_idx) + x;
+      }
+    }
+
+    return std::numeric_limits<double>::max();
+  }();
+
```

Comment:
```suggestion

// Insert deceleration point at stop distance without deceleration if use_constraints_for_decel is false
```

Reply:
Thanks for your suggestion! I added comments.

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/3785/commits/c0204c96a7126852db8c57a87fd77332e5a9ad7e